In [1]:
import pandas as pd
import numpy as np

In [44]:
corp=pd.read_csv("ner.csv",encoding="ISO-8859-1",error_bad_lines=False)

b'Skipping line 281837: expected 25 fields, saw 34\n'


In [45]:
corp.head()

Unnamed: 0     lemma next-lemma next-next-lemma next-next-pos  \
0           0  thousand         of        demonstr           NNS   
1           1        of   demonstr            have           VBP   
2           2  demonstr       have           march           VBN   
3           3      have      march         through            IN   
4           4     march    through          london           NNP   

  next-next-shape next-next-word next-pos next-shape      next-word  ...  \
0       lowercase  demonstrators       IN  lowercase             of  ...   
1       lowercase           have      NNS  lowercase  demonstrators  ...   
2       lowercase        marched      VBP  lowercase           have  ...   
3       lowercase        through      VBN  lowercase        marched  ...   
4     capitalized         London       IN  lowercase        through  ...   

  prev-prev-lemma prev-prev-pos prev-prev-shape prev-prev-word   prev-shape  \
0      __start2__    __START2__        wildcard     __START2__     wildcard   
1      __start1__    __START1__        wildcard     __START1__  capitalized   
2        thousand           NNS     capitalized      Thousands    lowercase   
3              of            IN       lowercase             of    lowercase   
4        demonstr           NNS       lowercase  demonstrators    lowercase   

       prev-word sentence_idx        shape           word tag  
0     __START1__          1.0  capitalized      Thousands   O  
1      Thousands          1.0    lowercase             of   O  
2             of          1.0    lowercase  demonstrators   O  
3  demonstrators          1.0    lowercase           have   O  
4           have          1.0    lowercase        marched   O  

[5 rows x 25 columns]

In [2]:
data=pd.read_csv("ner_dataset.csv",encoding="ISO-8859-1",error_bad_lines=False)

In [3]:
data.head(5)

Sentence #           Word  POS Tag
0  Sentence: 1      Thousands  NNS   O
1          NaN             of   IN   O
2          NaN  demonstrators  NNS   O
3          NaN           have  VBP   O
4          NaN        marched  VBN   O

In [4]:
data.shape

(1048575, 4)

In [21]:
corp.shape

NameError: name 'corp' is not defined

### Performing NER with sklearn

In [5]:
data.isna().sum()

Sentence #    1000616
Word                0
POS                 0
Tag                 0
dtype: int64

In [6]:
#eliminate NaN in first column
data = data.fillna(method='ffill')

In [15]:
#big data , cannot use it all
data=data[:10000]

In [16]:
x=data.drop(columns=["Sentence #","Tag"])

In [17]:
y=data["Tag"].values

In [18]:
classes=np.unique(data["Tag"])

In [19]:
#classes
classes=classes.tolist()

In [20]:
from sklearn.feature_extraction import DictVectorizer

In [21]:
vectorizer=DictVectorizer(sparse=False)
x = vectorizer.fit_transform(x.to_dict('records'))

In [22]:
from sklearn.model_selection import train_test_split

In [23]:
X_train,X_test,y_train,y_test=train_test_split(x,y,test_size=0.33,random_state=0)

In [24]:
X_train.shape,X_test.shape

((6700, 2785), (3300, 2785))

#### Out-of-core Algorithms
Here I use out of core algorithms because the data is too large and can't fit in my computer's memory at once

##### 1- Perceptron

In [25]:
from sklearn.linear_model import Perceptron
#Perceptron is like a NN with no hidden Layer
#N_Jobs = -1 means using all processors
per = Perceptron(verbose=10, n_jobs=-1, max_iter=5)
#partial_fit ==> OOC algorithms
per.partial_fit(X_train, y_train, classes)

[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done   1 tasks      | elapsed:    0.0s
[Parallel(n_jobs=-1)]: Done   2 out of  17 | elapsed:    0.0s remaining:    1.0s
[Parallel(n_jobs=-1)]: Done   4 out of  17 | elapsed:    0.1s remaining:    0.5s


-- Epoch 1
-- Epoch 1
-- Epoch 1
-- Epoch 1-- Epoch 1
-- Epoch 1-- Epoch 1-- Epoch 1


-- Epoch 1

-- Epoch 1-- Epoch 1Norm: 6.48, NNZs: 36, Bias: -2.000000, T: 6700, Avg. loss: 0.005075

-- Epoch 1
Total training time: 0.11 seconds.

Norm: 19.70, NNZs: 209, Bias: -2.000000, T: 6700, Avg. loss: 0.038358
Total training time: 0.10 seconds.
-- Epoch 1Norm: 4.47, NNZs: 17, Bias: -1.000000, T: 6700, Avg. loss: 0.001791

Total training time: 0.12 seconds.
-- Epoch 1
-- Epoch 1
Norm: 16.12, NNZs: 159, Bias: -3.000000, T: 6700, Avg. loss: 0.036119
Total training time: 0.14 seconds.
Norm: 4.24, NNZs: 10, Bias: -2.000000, T: 6700, Avg. loss: 0.001045
Total training time: 0.14 seconds.
-- Epoch 1
-- Epoch 1
Norm: 12.96, NNZs: 82, Bias: -1.000000, T: 6700, Avg. loss: 0.010448
Total training time: 0.13 seconds.
Norm: 16.06, NNZs: 151, Bias: -3.000000, T: 6700, Avg. loss: 0.023284
Total training time: 0.16 seconds.
Norm: 4.69, NNZs: 22, Bias: -2.000000, T: 6700, Avg. loss: 0.002985
Total training ti

[Parallel(n_jobs=-1)]: Done   6 out of  17 | elapsed:    0.1s remaining:    0.3s
[Parallel(n_jobs=-1)]: Done   8 out of  17 | elapsed:    0.1s remaining:    0.1s
[Parallel(n_jobs=-1)]: Done  10 out of  17 | elapsed:    0.1s remaining:    0.1s
[Parallel(n_jobs=-1)]: Done  12 out of  17 | elapsed:    0.1s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  14 out of  17 | elapsed:    0.1s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  17 out of  17 | elapsed:    0.2s finished


Perceptron(alpha=0.0001, class_weight=None, early_stopping=False, eta0=1.0,
           fit_intercept=True, max_iter=5, n_iter_no_change=5, n_jobs=-1,
           penalty=None, random_state=0, shuffle=True, tol=0.001,
           validation_fraction=0.1, verbose=10, warm_start=False)

We are going to eliminate the O Class since it's the most common tag because it will make our results look much better than what they actually are

In [26]:
new_classes=classes.copy()
new_classes.pop()

'O'

In [27]:
new_classes

['B-art',
 'B-eve',
 'B-geo',
 'B-gpe',
 'B-nat',
 'B-org',
 'B-per',
 'B-tim',
 'I-art',
 'I-eve',
 'I-geo',
 'I-gpe',
 'I-nat',
 'I-org',
 'I-per',
 'I-tim']

In [28]:
from sklearn.metrics import classification_report

In [29]:
print(classification_report(y_pred=per.predict(X_test),y_true=y_test,labels=new_classes))

              precision    recall  f1-score   support

       B-art       0.04      1.00      0.08         9
       B-eve       0.14      0.33      0.20         3
       B-geo       1.00      0.03      0.06        69
       B-gpe       0.69      0.62      0.65       102
       B-nat       0.00      0.00      0.00         0
       B-org       0.96      0.37      0.53        63
       B-per       0.95      0.46      0.62        41
       B-tim       0.64      0.96      0.77        52
       I-art       0.00      0.00      0.00        10
       I-eve       0.00      0.00      0.00         3
       I-geo       0.00      0.00      0.00        11
       I-gpe       1.00      0.17      0.29         6
       I-nat       0.00      0.00      0.00         1
       I-org       0.74      0.43      0.54        47
       I-per       0.88      0.21      0.34        66
       I-tim       0.00      0.00      0.00         4

   micro avg       0.42      0.41      0.42       487
   macro avg       0.44   

C:\Users\moham\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1437: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\Users\moham\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1439: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples.
  'recall', 'true', average, warn_for)


Here we clearly see that our classifier hasn't been able to correctly separate all classes. The best classified class is B-tim with f1-score=0.77 , precision 0.64 and recall 0.96. In this project we are interested to get all entities and classify them well that's why a good model must have a good Precision,Recall and f1-score.

##### 2- Stochastic Gradient Descent

Difference between Stochastic GD and GD :
In both gradient descent (GD) and stochastic gradient descent (SGD), you update a set of parameters in an iterative manner to minimize an error function. While in GD, you have to run through ALL the samples in your training set to do a single update for a parameter in a particular iteration, in SGD, on the other hand, you use ONLY ONE or SUBSET (Mini-Batch SGD) of training sample from your training set to do the update for a parameter in a particular iteration.

In [30]:
from sklearn.linear_model import SGDClassifier
sgd = SGDClassifier()
sgd.partial_fit(X_train, y_train, classes)

SGDClassifier(alpha=0.0001, average=False, class_weight=None,
              early_stopping=False, epsilon=0.1, eta0=0.0, fit_intercept=True,
              l1_ratio=0.15, learning_rate='optimal', loss='hinge',
              max_iter=1000, n_iter_no_change=5, n_jobs=None, penalty='l2',
              power_t=0.5, random_state=None, shuffle=True, tol=0.001,
              validation_fraction=0.1, verbose=0, warm_start=False)

In [31]:
print(classification_report(y_pred=sgd.predict(X_test), y_true=y_test, labels=new_classes))

              precision    recall  f1-score   support

       B-art       1.00      0.11      0.20         9
       B-eve       0.00      0.00      0.00         3
       B-geo       0.25      0.88      0.39        69
       B-gpe       0.87      0.54      0.67       102
       B-nat       0.00      0.00      0.00         0
       B-org       0.88      0.33      0.48        63
       B-per       0.61      0.61      0.61        41
       B-tim       1.00      0.81      0.89        52
       I-art       0.00      0.00      0.00        10
       I-eve       0.50      0.33      0.40         3
       I-geo       0.00      0.00      0.00        11
       I-gpe       0.75      0.50      0.60         6
       I-nat       0.00      0.00      0.00         1
       I-org       0.89      0.17      0.29        47
       I-per       0.86      0.18      0.30        66
       I-tim       0.25      0.25      0.25         4

   micro avg       0.51      0.47      0.49       487
   macro avg       0.49   

C:\Users\moham\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1437: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\Users\moham\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1439: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples.
  'recall', 'true', average, warn_for)


##### 3- Naive Bayes classifier for multinomial models

In [32]:
from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB(alpha=0.01)
nb.partial_fit(X_train, y_train, classes)

MultinomialNB(alpha=0.01, class_prior=None, fit_prior=True)

In [37]:
print(classification_report(y_pred=nb.predict(X_test), y_true=y_test, labels=new_classes))

              precision    recall  f1-score   support

       B-art       1.00      0.22      0.36         9
       B-eve       0.00      0.00      0.00         3
       B-geo       0.28      0.77      0.41        69
       B-gpe       0.45      0.72      0.55       102
       B-nat       0.00      0.00      0.00         0
       B-org       0.69      0.56      0.61        63
       B-per       0.81      0.54      0.65        41
       B-tim       0.96      0.87      0.91        52
       I-art       0.00      0.00      0.00        10
       I-eve       0.50      0.33      0.40         3
       I-geo       0.75      0.27      0.40        11
       I-gpe       0.29      0.33      0.31         6
       I-nat       0.00      0.00      0.00         1
       I-org       0.67      0.34      0.45        47
       I-per       0.78      0.38      0.51        66
       I-tim       0.09      0.50      0.15         4

   micro avg       0.49      0.57      0.53       487
   macro avg       0.45   

C:\Users\moham\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1437: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\Users\moham\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1439: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples.
  'recall', 'true', average, warn_for)


##### 4- Passive Aggressive Classifier

Passive: if correct classification, keep the model; Aggressive: if incorrect classification, update to adjust to this misclassified example, means that if the document is correctly classified, we keep in mind that it doesn't contain much information for updating.

In [36]:
from sklearn.linear_model import PassiveAggressiveClassifier
pac =PassiveAggressiveClassifier()
pac.partial_fit(X_train, y_train, classes)

PassiveAggressiveClassifier(C=1.0, average=False, class_weight=None,
                            early_stopping=False, fit_intercept=True,
                            loss='hinge', max_iter=1000, n_iter_no_change=5,
                            n_jobs=None, random_state=None, shuffle=True,
                            tol=0.001, validation_fraction=0.1, verbose=0,
                            warm_start=False)

In [39]:
print(classification_report(y_pred=pac.predict(X_test), y_true=y_test, labels=new_classes))

              precision    recall  f1-score   support

       B-art       0.00      0.00      0.00         9
       B-eve       0.00      0.00      0.00         3
       B-geo       0.23      0.90      0.37        69
       B-gpe       0.92      0.55      0.69       102
       B-nat       0.00      0.00      0.00         0
       B-org       0.91      0.16      0.27        63
       B-per       1.00      0.41      0.59        41
       B-tim       0.90      0.88      0.89        52
       I-art       0.00      0.00      0.00        10
       I-eve       0.03      0.33      0.05         3
       I-geo       0.00      0.00      0.00        11
       I-gpe       1.00      0.17      0.29         6
       I-nat       0.00      0.00      0.00         1
       I-org       0.62      0.11      0.18        47
       I-per       0.88      0.11      0.19        66
       I-tim       0.00      0.00      0.00         4

   micro avg       0.44      0.42      0.43       487
   macro avg       0.41   

C:\Users\moham\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1437: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\Users\moham\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1439: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples.
  'recall', 'true', average, warn_for)


Based on f1-score we can conclude that the Naive Bayes Multinomial classifier is the best one with an average of 0.53 
(à suivre)